### ISSUE TITLE
Improve JS outline/breadcrumb display for arrow functions

### ISSUE THREAD

**Issue by ryaninvents** [Microsoft/vscode#58367]
After [this Twitter discussion](https://twitter.com/ryaninvents/status/1038092296103837696), it seems that there's interest in improving the outline/breadcrumb display for arrow functions.

My proposal: if the arrow function is an argument in a function call, use the name of the called function along with any string literals to visually identify the otherwise unnamed function.

For example:

```js
// Example code
document.addEventListener('click', () => {
  // This would ordinarily display as "<function>"
})

/* Proposed outline display

> addEventListener('click', ...)

*/
```

I took a quick look at the code but wasn't sure how the tokens are generated or used to form the outline, but I wrote [a proof-of-concept](http://astexplorer.net/#/gist/2514a0b1d3cf3490e9119b48e19c3d29/8ee2e7dd79042ee2548d013b4ab91c039f41ca3c) using Babel to at least give a sense of my idea.

### Before:

![](https://pbs.twimg.com/media/DmgBNLqXgAEvFjb.jpg)

### After:

![](https://pbs.twimg.com/media/DmgLSPXUcAA-92_.png)

**Comment by ryaninvents** [comment]
I spent a bit of time investigating, and my best guess is that I'd have to insert a lookback of some sort into [this monster RegEx](https://github.com/Microsoft/TypeScript-TmLanguage/blob/c90418c57af09ddc7bd6827895052a6276ef8a9f/TypeScriptReact.tmLanguage#L618), so that if it's an arrow function it'll try and read the preceding function call and string arg.

**This is a total guess though**; if someone more familiar with the codebase knows where to look then I'd be happy to listen.

**Comment by mjbvz** [comment]
This issue was moved to Microsoft/TypeScript#27051

**Comment by connectdotz** [comment]
hmmm... I am on vscode 1.28.2 and typescript 3.1.3, and my outline still display `<function>` for the test blocks, exactly like @ryaninvents mentioned. Given this issue is closed, I was wondering why it is not working yet...

I did a little bit of digging, and for those who are also wondering, looks like the typescript fix ([Microsoft/TypeScript#27063](https://github.com/Microsoft/TypeScript/pull/27063)) is scheduled for typescript [3.2](https://github.com/Microsoft/TypeScript/milestone/77?closed=1), so at least until then, we won't see the improvement... probably best to keep this issue open until the fix is confirmed on the vscode side, in case there is some work needed here as well...

**Comment by ryaninvents** [comment]
Yeah, it was closed in favor of the [upstream issue](https://github.com/Microsoft/TypeScript/issues/27051). Unfortunately, even when that fix lands in TS3.2 I don't think it will affect me, since I use the Babel grammar instead.

**Comment by mjbvz** [comment]
The fix in only in TypeScript 3.2 which we won't pick up until November. To test it out, install `typescript@next` by following [these instructions](https://code.visualstudio.com/Docs/languages/typescript#_using-newer-typescript-versions)

@ryaninvents The online comes from the typescript language service, not from a grammar
